File: src/main/java/com/personal/array/InterpolationSearch.java
package com.personal.array;

public class InterpolationSearch {

    static int interpolation_search(int arr[], int size, int key) {
        int low = 0;
        int high = size - 1;
        int mid;

        while (arr[high] != arr[low] && key >= arr[low] && key <= arr[high]) {
            mid = low + ((key - arr[low]) * (high - low) / (arr[high] - arr[low]));

            if (arr[mid] < key)
                low = mid + 1;
            else if (key < arr[mid])
                high = mid - 1;
            else
                return mid;
        }

        if (key == arr[low])
            return low;
        else
            return -1;
    }


    public static void main(String... args) {
        int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};

        int index = interpolation_search(arr, 9, 8);
        System.out.printf("Index position for Item: %d is %d", arr[index], index);

    }
}

File: src/main/python/basic/interpolation_search.py
#!/usr/bin/python


def interpolation_search(array, key):
    print("Searching key {}".format(key))
    low = 0
    high = len(array) - 1

    while low <= high and array[low] <= key <= array[high]:
        pos = low + int(((high - low) / (array[high] - array[low])) * (key - array[low]))

        if array[pos] == key:
            return pos
        elif key > array[pos]:
            low = pos + 1
        else:
            high = pos - 1
    return -1


if __name__ == '__main__':
    input_array = [2, 5, 9, 15, 20, 26, 30, 40, 51, 56, 60]
    position = interpolation_search(input_array, 20)
    if position == -1:
        print("Element not found")
    else:
        print("Element found at position {}".format(position))
